libraries/Microsoft.Bot.PublishValidation/Helpers/BotValidatorHelper.cs

```diff
@@ -0,0 +1,377 @@
+                        messages = messages.Append(message);
+                        validationResult = false;
+                    }
+                }
+                
+                // Check if the .bot file has a Luis Key
+                if (options.RequireLuisKey)
+                {
+                    if (!BotValidatorHelper.ValidateLuisKey(botConfiguration))
+                    {
+                        errorMsg = "The .bot file does not have a Luis Key.";
+                        message = new NotificationMessage(errorMsg, NotificationMessageTypes.Error);
+                        messages = messages.Append(message);
+                        validationResult = false;
+                    }
+                }
+
+                // Check if the .bot file has a qna Key
+                if (options.RequireQnAMakerKey)
+                {
+                    if (!BotValidatorHelper.ValidateQnAKey(botConfiguration))
+                    {
+                        errorMsg = "The .bot file does not have a QnA Key.";
+                        message = new NotificationMessage(errorMsg, NotificationMessageTypes.Error);
+                        messages = messages.Append(message);
+                        validationResult = false;
+                    }
+                }
+
+                return validationResult;
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
+        }
+
+        /// <summary>
+        /// Validate the specified endpoints of a .bot file, according to if they are required or not.
+        /// </summary>
+        /// <param name="botConfiguration"></param>
+        /// <param name="specifiedEndpoints"></param>
+        /// <param name="required"></param>
+        /// <missingEndpoints name="missingEndpoints">List of the missing endpoints in the bot file. If its null/empty and the methods returns FALSE, it means that there isn't ANY endpoint in the file.</missingEndpoints>
+        /// <returns></returns>
+        private static bool ValidateEndpoints(BotConfiguration botConfiguration, IEnumerable<string> endpoints, bool required, out string missingEndpoints)
+        {
+            var missingEndpointsList = new List<string>();
+
+            // Get all the endpoint types in the service list
+            var botEndpoints =
+                BotValidatorHelper.GetBotServices(botConfiguration, BotServiceType.endpoint);
+
+            // If there isn't any endpoint, returns false
+            if (botEndpoints == null)
+            {
+                missingEndpoints = string.Empty;
+                return false;
+            }
+
+            // Checks that all the specified endpoints are/aren't in the bot file
+            foreach (var endpoint in endpoints)
+            {
+                if (botEndpoints.Any(ep => ((EndpointService)ep).Name.Trim() == endpoint.Trim()) != required)
+                {
```

Comment:
I can't follow this if statements - too complex. Can you simplify?

Reply:
We've divided that line into a few more and also added comments.